### PR TITLE
[SPARK-44154] Implement bitmap functions

### PR DESCRIPTION
**Pull request by gene-db**
### What changes were proposed in this pull request?

Implemented bitmap functions. The functions are:
- `bitmap_bucket_number()`: returns the bucket number for a given input number
- `bitmap_bit_position()`: returns bit position for a given input number
- `bitmap_count()`: returns the number of set bits from an input bitmap
- `bitmap_construct_agg()`: aggregation function that aggregates input bit positions, and creates a bitmap
- `bitmap_or_agg()`: aggregation function that performs a bitwise OR on all the input bitmaps

### Why are the changes needed?

These functions can be used to count distinct values for integer columns. For example:

```sql
SELECT sum(cnt) FROM (
  SELECT bitmap_bucket_number(c), bitmap_count(bitmap_construct_agg(bitmap_bit_position(c))) cnt
  FROM table GROUP BY 1
)
```

is equivalent to:

```sql
SELECT count(distinct c) FROM table
```

### Does this PR introduce _any_ user-facing change?

Yes. After this PR, these functions are usable in queries.

### How was this patch tested?

New tests were added.

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/bitmapExpressions.scala [hunk 1]
```diff
@@ -0,0 +1,318 @@
...
+      > SELECT _FUNC_(123);
+       1
+      > SELECT _FUNC_(0);
+       0
+  """,
+  since = "3.5.0",
+  group = "misc_funcs"
+)
+case class BitmapBucketNumber(child: Expression)
+  extends UnaryExpression with ImplicitCastInputTypes with NullIntolerant with CodegenFallback {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/bitmapExpressions.scala:56]: It's more recommended to implement an expression with a static java method. See `AesEncrypt` as an example.

Comment (gene-db) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/bitmapExpressions.scala]: I updated these expressions to use `StaticInvoke()`.

**Review by gene-db — COMMENTED** [review]
@cloud-fan Thanks! I updated the PR with your suggestion.

File: sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/BitmapExpressionUtils.java [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.catalyst.expressions;
+
+/**
+ * A utility class for constructing bitmap expressions.
+ */
+public class BitmapExpressionUtils {
+  /** Number of bytes in a bitmap. */
```
Comment (cloud-fan) [inline sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/BitmapExpressionUtils.java:24]: why is it a fixed number?

Comment (gene-db) [reply sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/BitmapExpressionUtils.java:24]: That is the size that Oracle and Snowflake are using, and it looks like that how these bitmaps are sized. It doesn't look like there is a way to change the size.

**Review by cloud-fan — APPROVED** [review]

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Pull request closed without merging**